File: core/causality/state_transition_engine.py
from __future__ import annotations

from typing import Any, Dict, List


def build_state_transitions(
    events: List[Dict[str, Any]],
) -> Dict[str, Any]:
    transitions: List[Dict[str, Any]] = []

    for index in range(1, len(events)):
        prev = events[index - 1]
        curr = events[index]
        transitions.append({
            "from_state": str(prev.get("state", prev.get("type", ""))),
            "to_state": str(curr.get("state", curr.get("type", ""))),
            "runtime": str(curr.get("runtime", "")),
            "step": index,
        })

    return {
        "transitions": transitions,
        "count": len(transitions),
        "bounded": True,
    }

File: src/causality/stateTransitionEngine.ts
/**
 * Converted from Python: core/causality/state_transition_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function buildStateTransitions(events: any): any {
  var transitions: any[] = [];
  var index: any;
  for (index = 1; index < py.len(events); index++) {
    var prev: any = py.at(events, py.sub(index, 1));
    var curr: any = py.at(events, index);
    py.listAppend(transitions, {"from_state": py.toStr(py.get(prev, "state", py.get(prev, "type", ""))), "to_state": py.toStr(py.get(curr, "state", py.get(curr, "type", ""))), "runtime": py.toStr(py.get(curr, "runtime", "")), "step": index});
  }
  return {"transitions": transitions, "count": py.len(transitions), "bounded": true};
}
